enhancements/automated-service-ca-rotation.md

```diff
@@ -0,0 +1,271 @@
+  current CA to determine whether the serving cert was issued by the current CA.
+  - If the serving cert is determined not to have been issued by the current CA,
+    it needs to be regenerated.
+  - Previously, `Subject.CommonName` of a serving cert was compared with
+    `Issuer.CommonName` of the CA. Since ensuring trust between old and new CAs
+    requires that `Subject.CommonName` be reused, this comparison will no longer
+    be sufficient to determine lineage.
+  - This requires that all new certificates - whether CA or entity - have
+    `AuthorityKeyId` and `SubjectKeyId` set.
+  - In addition to ensuring regeneration when `AuthorityKeyId` and `SubjectKeyId`
+    are set, this comparison schema will also ensure regeneration of serving
+    certs without `SubjectKeyId` set (due to being generated by a CA without
+    `AuthorityKeyId` set) since an empty `SubjectKeyId` will not match a
+    populated `AuthorityKeyId`.
+
+- Provide an indication that rotation has taken place via an event.
+
+- Allow rotation to be forced when
+  `ServiceCA.Spec.ForceCARotation.Reason` is set to a non-empty value
+  that does not match the value for the
+  `service.beta.openshift.io/forced-rotation-reason` annotation on the
+  signing secret.
+  - Setting an annotation on the signing secret ensures that a forced
+    rotation is triggered at most once per reason.
+  - If a forced rotation and expiry-triggered rotation coincide, only
+    the forced rotation should be performed.
+
+- Allow the duration of the grace period for continuing to trust the
+  previous CA to be overridden when forcing rotation by setting
+  `ServiceCA.Spec.ForceCARotation.GraceDuration`.
+  - If set, the expiry of the intermediate certs bridging trust
+    between the pre-rotation serving certs and CA bundle will be set
+    to the time of rotation plus the provided duration.
+
+- Allow trust to be maintained across multiple rotations by setting
+  `ServiceCA.Spec.ForceCARotation.MaintainPriorTrust: true` (primarily
+  to support e2e testing). If a rotation was previously performed from
+  CA A to CA B, a forced rotation to CA C would exhibit the following
+  behavior (A<sub>b</sub> -> the public key of CA A signed by CA B):
+  - If `MaintainPriorTrust: false` (the default):
+    - The set of intermediate certs appended to service certs issued
+      by the new CA consists of the new CA's primary certificate
+      issued by the previous CA (e.g. [C<sub>b</sub>]).
+    - The new CA bundle includes the primary certificate of the
+      previous CA signed by the new CA and the primary certificate of
+      the new CA (e.g. [B<sub>c</sub>,C<sub>c</sub>]).
+  - If `MaintainPriorTrust: true`:
+    - The set of intermediate certs appended to service certs issued
+      by the new CA consists of the previous set of intermediate certs
+      in addition to the new CA's primary certificate reissued by the
+      previous CA (e.g. [B<sub>a</sub>,C<sub>b</sub>]).
```

Comment:
this means the number of intermediate certs grows linearly with rotation. More importantly, the length of the chain to verify the new cert with an old CA grows the same. Is there an upper bound that the TLS stack allows?

Reply:
https://security.stackexchange.com/questions/117169/how-long-can-x-509-certificate-chains-be tells that there is no limit on the chain length by default, only through some extension, that we probably do not set anyway.

Reply:
I'm not aware of a hard limit, but since the set of intermediate certs would be sent to the client when negotiating a new connection it could impose undesired overhead at some point. Again, if we don't have a use-case for this feature I'm not sure we should be implementing it.